src/librustc/ty/sty.rs

```diff
@@ -475,18 +470,21 @@ impl<'tcx> GeneratorSubsts<'tcx> {
     /// It contains a tuple of all the types that could end up on a generator frame.
     /// The state transformation MIR pass may only produce layouts which mention types
     /// in this tuple. Upvars are not counted here.
-    pub fn witness(self, def_id: DefId, tcx: TyCtxt<'_>) -> Ty<'tcx> {
-        self.split(def_id, tcx).witness
+    pub fn witness(self) -> Ty<'tcx> {
+        self.split().witness
+    }
+
+    pub fn tupled_upvars_ty(self) -> Ty<'tcx> {
+        self.split().tupled_upvars_ty
     }
 
     #[inline]
-    pub fn upvar_tys(
-        self,
-        def_id: DefId,
-        tcx: TyCtxt<'_>,
-    ) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
-        let SplitGeneratorSubsts { upvar_kinds, .. } = self.split(def_id, tcx);
-        upvar_kinds.iter().map(|t| {
+    pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
+        let upvars = match self.split().tupled_upvars_ty.kind {
+            Tuple(upvars) => upvars,
+            _ => bug!("upvars should be tupled"),
+        };
+        upvars.iter().map(|t| {
```

Comment:
```suggestion
        self.split().tupled_upvars_ty.tuple_fields().map(|t| {
```

Reply:
Considered that and `expect_ty`, worried it was less informative if it failed.

Reply:
you can slap `#[track_caller]` on the two functions, that should mitigate that, right?